### ISSUE TITLE
Training Aux-head and main heads

### ISSUE THREAD

**Issue by paulhigazi** [xu-ji/IIC#55]
Just read your paper, amazing work ! I am just about to finish my keras implementation, but have a small understanding problem. When, and in which order do i train the different heads ?

Am i going to:

1) freeze the weigths of all subheads of aux and main , and than just backprop over one subhead for a whole epoch ?

or 

2) calculate the loss over all main heads for one epoch, sum the loss of them up and backprop than, next epoch do the same for aux ?

or sth completely different?

Sorry i tried to reverse-engineer your pytorch code but i am not into pytorch, and the forks also didnt help me. thanks a lot for your time!

**Comment by xu-ji** [comment]
Closer to 2:

```
heads = [main, aux] # or other order
for ep in epochs:
  for head in heads:
    for batch in data(head):
      compute loss for all sub-heads in head, average over sub-heads, and do full backprop
```
 You could also try backprop from a joint loss on both heads at the same time without switching between them, if you have enough GPU memory (we didn't use this).
